[LinearSolver] Untangle linear solver and linear system

alxbilger opened the pull request:
In #2777, the matrix assembly was removed from the linear solver. Instead, the assembly is done in a dedicated component, a linear system. However, linear solvers still kept an API related to the matrix assembly. This PR removes this API for a clearer one, where both solvers and systems have their own responsibilities.

With this work, we will be able to assemble different systems while keeping the same linear solvers. For example, an augmented system containing also lagrangian constraints.

An interesting consequence is that a linear solver is now agnostic to the simulated physic: no mass factor, stiffness factor, damping factor. The assembly of the mass, damping and stiffness is now totally in the linear system. We can now imagine a linear system assembling different types of physics, while keeping the same linear solvers.

I took the opportunity to clean a few things in the preconditioners.

Additions:
- A new Data `authorizeAssembly` in `BaseMatrixLinearSystem`. This Data is hidden to the user. This Data exists to use the DDGNode system, in particular in the asynchronous solver.
- Similarly, a new Data `factorizationInvalidation` is introduced in `MatrixLinearSolver`. This Data is linked to the Data `matrixChanged`, also newly introduced, and serves as an observer for matrix changes. That way, the solver decompose the matrix only if a change in the matrix occurred.
- `CompositeLinearSystem` is now templated also on `GraphScatteredMatrix`.
- `PreconditionedMatrixFreeSystem` is a new component of the categories of the linear systems. It is intended to work with `PCGLinearSolver`. The specificity of this linear system is the link toward a the linear system of the preconditioner. It has a Data `assemblingRate`, allowing to assemble the matrix of the preconditioner only at a given rate. It is templated only on `GraphScatteredMatrix` since it works only with `PCGLinearSolver`.

Changes:
- Some types involving scalar types are changed from `double` to `Real`
- When there is no linear system in the scene, the linear solver instantiates automatically a linear system component. Before it was added as a slave, mainly for UX reasons. However, slaves are not visited by the visitors. Therefore, linear systems are no longer a slave when instantiated automatically.
- The Data `update_step` is removed in `PCGLinearSolver`. Since the solvers are no longer responsible to assemble the matrices, this Data is no longer used. Instead, the equivalent of this Data is now in `PreconditionedMatrixFreeSystem`.
- The assembly in the `WarpPreconditioner` is now transferred to `RotationMatrixSystem`. Similarly to `PreconditionedMatrixFreeSystem`, it has also a Data `assemblingRate`.
- The rotation finder is now an explicit link in `RotationMatrixSystem`. Before it was found implicitly with an index in `WarpPreconditioner`.

To do:

- [x] Fix warp preconditioner
- [x] Proper deprecation
- [ ] Review of the constraints, including the preconditioners
- [x] Fix `MechanicalMultiVectorPeqBaseVectorVisitor` in `MatrixLinearSolver`


[ci-depends-on https://github.com/sofa-framework/SofaPython3/pull/528]

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

alxbilger committed:
Start removing assembly function from linear solvers

alxbilger committed:
move internal functions to protected

alxbilger commented:
[ci-build][with-all-tests]

alxbilger committed:
cache factorization + cache invalidation

alxbilger committed:
remove more "assembly" functions

alxbilger committed:
fix AsyncSparseLDLSolver

alxbilger committed:
add missing templates to the factory

alxbilger committed:
CompositeLinearSystem can support multiple types of linear systems

alxbilger committed:
introduce preconditioned systems

alxbilger committed:
a bit of cleaning

alxbilger committed:
convert double to Real

alxbilger committed:
add an error if wrong type

alxbilger committed:
Restoration of the parameter controlling the assembly rate of the preconditioned matrix. It was the Data 'update_step' in PCGLinearSolver, but now it's the Data 'assemblingRate' in PreconditionedMatrixFreeSystem.

Benchmark of examples\Component\LinearSolver\Preconditioner\FEMBAR_PCG_JacobiPreconditioner.scn on 1000 time steps with variation of `assemblingRate`:

1: 700.128 FPS
2: 968.381 FPS
3: 1074.87 FPS
4: 1156.56 FPS
10: 1383.11 FPS
20: 1478.05 FPS
30: 1561.64 FPS
60: 1563.85 FPS

alxbilger committed:
restore removed Data with a deprecation layer

alxbilger committed:
restore initial parameter for regression tests

alxbilger committed:
fix PrecomputedLinearSolver

bakpaul commented on Sofa/framework/Core/src/sofa/core/behavior/LinearSolver.h:
```diff
@@ -118,19 +96,18 @@ class SOFA_CORE_API LinearSolver : public BaseLinearSolver
-    virtual void applyConstraintForce(const sofa::core::ConstraintParams* /*cparams*/,sofa::core::MultiVecDerivId /*dx*/, const linearalgebra::BaseVector* /*f*/) {
-        msg_error() << "applyConstraintForce has not been implemented.";
-    }
+    virtual void applyConstraintForce(const sofa::core::ConstraintParams* /*cparams*/,
```
I know, not in the scope of this PR, but why do we need this ? Isn't addMinvJt enough ?

bakpaul commented:
Ok so now the only purpose of a linear solver is to solve inverse linear problem given a LinearSystem that is responsible of defining the RHS and the LHS ? 

If the solver is a direct one, let's says a Cholesky solver, where is the decomposition supposed to be stored ? If the Solver is agnostic to the system, then it should not keep track of this decomposition no ? If in the future we would like to have many different LinearSystem using the same solver then they should somehow keep this info. Or maybe an intermediate map in the solver ?

alxbilger commented:
> Ok so now the only purpose of a linear solver is to solve inverse linear problem given a LinearSystem that is responsible of defining the RHS and the LHS ?
> 
> If the solver is a direct one, let's says a Cholesky solver, where is the decomposition supposed to be stored ? If the Solver is agnostic to the system, then it should not keep track of this decomposition no ? If in the future we would like to have many different LinearSystem using the same solver then they should somehow keep this info. Or maybe an intermediate map in the solver ?

@bakpaul The linear solver is still bound to a single linear system through its Link. It think it still makes sense to have the decomposition in the solver because of that. 

The intention with this PR was only to remove the assembly part in the API of the solver.

alxbilger committed:
mark WarpPreconditioner destructor as `override`

alxbilger committed:
minor cleaning

alxbilger committed:
simplify `checkLinearSystem` in WarpPreconditioner implementation

alxbilger committed:
add license header to RotationMatrixSystem header file

alxbilger committed:
fix typo in registration description of RotationMatrixSystem

alxbilger committed:
modernize code style

alxbilger committed:
add missing `RotationMatrixSystem.inl` to CMakeLists.txt

alxbilger committed:
replace `addSlave` with `addObject` in MatrixLinearSolver

This is to allow the added object to be initialized. Slaves are not initialized because visitors don't visit them. They are owned by their master, then it's up to their master to initialize them or not. There is no reason for a linear system to be a slave.

alxbilger committed:
add SOFA class macros for proper class hierarchy

alxbilger committed:
Missing space

alxbilger committed:
proper initialization of systems and solvers

alxbilger committed:
validate `l_linearSystem` type in `PCGLinearSolver`

alxbilger committed:
fix typo in error message of `PCGLinearSolver`

alxbilger committed:
validate `l_mainAssembledSystem` and handle missing or invalid linear systems in `RotationMatrixSystem`

alxbilger committed:
validate `l_linearSystem` type and ensure compatibility with `RotationMatrixSystem` in `WarpPreconditioner`

alxbilger committed:
don't link to itself

alxbilger committed:
refactor `RotationMatrixSystem` and `WarpPreconditioner` for clearer logic and enhanced matrix handling

alxbilger committed:
some comments

alxbilger committed:
proper deprecation

alxbilger committed:
update scene

alxbilger committed:
minor changes

alxbilger committed:
add `template` keyword to `get` and `getObjects` calls in `RotationMatrixSystem` for proper type resolution

alxbilger committed:
invert includes order to allow object factory to find the component module

alxbilger committed:
apply changes

alxbilger committed:
deprecation

alxbilger commented:
[ci-build][with-all-tests]

alxbilger committed:
forgot a file

alxbilger committed:
simplify `checkLinearSystem` logic using `doCheckLinearSystem`

alxbilger committed:
add components that are created implicitly in the scene

alxbilger committed:
add a description to a class member

alxbilger committed:
see if regression test is happy

alxbilger commented:
⚠️ The example scene `FEMBAR_PCG_PrecomputedWarpPreconditioner.scn` does not move, but it is also the case in master.

fredroy commented on Sofa/framework/Core/src/sofa/core/behavior/BaseMatrixLinearSystem.cpp:
```diff
@@ -27,15 +27,22 @@ namespace sofa::core::behavior
 BaseMatrixLinearSystem::BaseMatrixLinearSystem()
 : Inherit1()
 , d_matrixSize(initData(&d_matrixSize, "matrixSize", "Size of the global matrix"))
+, d_authorizeAssembly(initData(&d_authorizeAssembly, true, "authorizeAssembly", "Allows to assemble the system matrix"))
```
```suggestion
, d_authorizeAssembly(initData(&d_authorizeAssembly, true, "authorizeAssembly", "Enable assembly of the system matrix"))
```
and maybe renaming the Data with d_enableAssembly (`authorize` seems ... too strong ?)

fredroy replied:
This applies to the other recurrences of this `d_authorizeAssembly` Data in other components

alxbilger replied:
done

alxbilger committed:
rename `d_authorizeAssembly` to `d_enableAssembly`

fredroy commented on Sofa/Component/LinearSolver/Direct/src/sofa/component/linearsolver/direct/AsyncSparseLDLSolver.inl:
```diff
@@ -29,44 +29,58 @@ namespace sofa::component::linearsolver::direct
 template <class TMatrix, class TVector, class TThreadManager>
-void AsyncSparseLDLSolver<TMatrix, TVector, TThreadManager>::init()
+AsyncSparseLDLSolver<TMatrix, TVector, TThreadManager>::AsyncSparseLDLSolver()
+    : d_enableAssembly(initData(&d_enableAssembly, true, "authorizeAssembly", "Allow assembly of the linear system"))
```
```suggestion
    : d_enableAssembly(initData(&d_enableAssembly, true, "enableAssembly", "Allow assembly of the linear system"))
```

alxbilger committed:
rename Data name

alxbilger committed:
fix call to MechanicalMultiVectorPeqBaseVectorVisitor but actually it is not called

alxbilger commented:
Impossible, gmock is not available in SOFA

fredroy commented:
[ci-build][with-all-tests][force-full-build]
